pkg/syncer/resource_syncer.go

```diff
@@ -322,89 +326,99 @@ func (r *resourceSyncer) ListResources() []runtime.Object {
+			// reverse transform function to handle this. Possible future work, for now bail.
+			r.log.Warningf("Unable to reconcile type %T - expected type %v", resource, resourceType)
 			return
 		}
 
-		resourceType := reflect.TypeOf(r.config.ResourceType)
+		metaObj := resourceUtil.MustToMeta(resource)
+		clusterID, found := getClusterIDLabel(resource)
+		ns := r.config.SourceNamespace
 
-		for _, resource := range resourceLister() {
-			if reflect.TypeOf(resource) != resourceType {
-				// This would happen if the custom transform function returned a different type. We would need a custom
-				// reverse transform function to handle this. Possible future work, for now bail.
-				r.log.Warningf("Unable to reconcile type %T - expected type %T", resource, resourceType)
-				return
+		switch r.config.Direction {
+		case None:
+			ns = metaObj.GetNamespace()
+		case RemoteToLocal:
+			if !found || clusterID == r.config.LocalClusterID {
+				continue
 			}
-
-			metaObj := resourceUtil.MustToMeta(resource)
-			clusterID, found := getClusterIDLabel(resource)
-			ns := r.config.SourceNamespace
-
-			switch r.config.Direction {
-			case None:
-				ns = metaObj.GetNamespace()
-			case RemoteToLocal:
-				if !found || clusterID == r.config.LocalClusterID {
-					continue
-				}
-			case LocalToRemote:
-				if clusterID != r.config.LocalClusterID {
-					continue
-				}
-
-				labels := metaObj.GetLabels()
-				delete(labels, federate.ClusterIDLabelKey)
-				metaObj.SetLabels(labels)
-
-				if ns == metav1.NamespaceAll {
-					ns = metaObj.GetLabels()[OrigNamespaceLabelKey]
-				}
+		case LocalToRemote:
+			if clusterID != r.config.LocalClusterID {
+				continue
 			}
 
-			if ns == "" {
-				r.log.Warningf("Unable to reconcile resource %s/%s - cannot determine originating namespace",
-					metaObj.GetNamespace(), metaObj.GetName())
-				continue
+			labels := metaObj.GetLabels()
+			delete(labels, federate.ClusterIDLabelKey)
+			metaObj.SetLabels(labels)
+
+			if ns == metav1.NamespaceAll {
+				ns = metaObj.GetLabels()[OrigNamespaceLabelKey]
 			}
+		}
 
-			metaObj.SetNamespace(ns)
+		if ns == "" {
+			r.log.Warningf("Unable to reconcile resource %s/%s - cannot determine originating namespace",
+				metaObj.GetNamespace(), metaObj.GetName())
+			continue
+		}
 
-			key, _ := cache.MetaNamespaceKeyFunc(resource)
+		metaObj.SetNamespace(ns)
 
-			_, exists, _ := r.store.GetByKey(key)
-			if exists {
-				continue
-			}
+		key, _ := cache.MetaNamespaceKeyFunc(resource)
 
-			obj := resourceUtil.MustToUnstructuredUsingScheme(resource, r.config.Scheme)
-			r.deleted.Store(key, obj)
-			r.workQueue.Enqueue(obj)
+		_, exists, _ := r.store.GetByKey(key)
+		if exists {
+			continue
 		}
-	}()
+
+		obj := resourceUtil.MustToUnstructuredUsingScheme(resource, r.config.Scheme)
+		r.deleted.Store(key, obj)
+		r.workQueue.Enqueue(obj)
+	}
+}
+
+func (r *resourceSyncer) runIfWaitForCacheSync(defaultReturn any, run func() any) any {
```

Comment:
I understand the construction, but I found “run if wait for cache sync” weird when reviewing this. `waitForCacheSyncThenRun` or `runIfCacheSynced` seem more accurate to me.